### PR TITLE
Implement `weakref` module from `__del__` only.

### PR DESCRIPTION
**Pull request by AJMansfield**
### Summary
This PR demonstrates an alternate concept to #18822 for a, implementation of the `weakref` module that relies more heavily on python-language code atop the `__del__` method plus a much simpler c-language "weak pointer" primitive.

This is built on #18005, though in theory any implementation of `__del__` that handles cyclic isolates correctly would also work.

### Testing
I've pulled in all tests from #18822 for comparison. Note that few of the tests pass as written, but for many of them the distinction is down to just sequential ordering. There's also one lingering exception-handling issue but I'm confident I'll be able to resolve it.

I've flat-enabled this feature on all ports for the sake of size comparison --- this is not intended to be merged as such.

### Trade-offs and Alternatives
This is an alternative proposal to #18822, aimed at a much smaller code size suitable for all ports, but likely with much worse runtime performance.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for posting this, it's very interesting to see an alternative to #18822.

I see that the unix coverage build is not running the weakref tests, I think it needs to have `weakref.py` added to `ports/unix/variants/coverage/manifest.py`.

> Note that few of the tests pass as written, but for many of them the distinction is down to just sequential ordering.

I tried to make sure the tests I added don't rely on sequence ordering, and for most of them they match exactly the CPython output (there's no corresponding .py.exp file).  Only a few of the tests actually test the order in which the GC collects objects.

Looking at your code, I'm not sure how the following case is handled, which seems to be failing the test currently in this PR: a `r=ref(obj)` is created, then `r=None` is executed which means the ref gets collected, then `obj=None` is executed, and the ref callback is NOT called.  I went to some lengths to make sure that was handled correctly, ie if a `ref` is collected then it cancels any weakref callbacks that it had registered.

**Comment by dpgeorge** [comment]
@AJMansfield I've taken a good look at this PR.  It's very clever how you've implemented the weakref callback scheme, whereby the object points via `__weakref__` to a `_weakref_entry` instance, and the latter is automatically collected when the object is collected.

From my understanding it works like this:
```
+-----------+                     +----------------+
| user      | -- .__weakref__ --> |                |
| object    |                     | _weakref_entry |
|           | <----- .next ------ |                |
+-----------+                     +----------------+
      ^                               /
       .                             /
        . weak pointer      .parent /
         .                         /
          .                       /
           .                     v
         +-------------------------+
         |          ref            |
         |                         |
         |  (has .__callback__)    |
         +-------------------------+
```
All arrows are strong pointers, except the dotted weak pointer from `ref` to the user object.

When the user object no longer has any other pointers pointing to it (ie only those in the diagram above remain) then the GC will collect it.  The GC will also collect `_weakref_entry` because the only pointers to it are from the user object, which is no longer alive.

When `_weakref_entry` gets collected, it has a `__del__` (implemented in Python) which then looks at `.parent` to find the `ref` object and call its callback.

There are three things with this approach which I think are currently problematic (although maybe I don't understand it properly):
1. `_weakref_entry.__del__` uses `micropython.schedule()` to schedule the `ref`'s callback to happen "later" (after the GC cycle is completed).  There is only a finite (small) queue of entries in the scheduler so if many objects are collected at the same time and they have callbacks, only the first few will run, the rest are lost.
2. CPython semantics are such that, if the `ref` instance is reclaimed (ie because the user no longer has any variable pointing to it) then the weakref callback is cancelled.  I don't think that can happen with the above scheme, because `ref` always has a strong pointer to it from `_weakref_entry`, which has a strong pointer from the user object.  So as long as the user object is alive, so is `ref`.  That means the callback will always run, even if the user tried to "delete" the `ref` by forgetting all its references.  (`finalize` should have this behaviour of always being alive, but not `ref`.)
3. Because `_weakref_entry` is a separate object to the user object (and to `ref`) it could be that it has a stray pointer pointing to it (eg some random data in an array with just the right value) and hence does not get collected when the user object gets collected (this is because MicroPython's GC is not precise, it's conservative).  If such a thing happens (which it will eventually if the app uses lots of weak refs over a long period of time and has lots of other data on the heap) then (a) the callback is not executed as expected, and (2) `ref` still thinks the user object is alive and calling `ref` will return a pointer to an invalid location in the heap.

Points (1) and (2) can probably be fixed with extra code (eg preallocating a linked list of callbacks that need to run, and using extra weakrefs to point from `_weakref_entry` back to `ref`), but I don't know how to solve point (3).

@AJMansfield what do you think of this analysis, is there something I missed?

**Comment by AJMansfield** [comment]
@dpgeorge That's all correct, including the three limitations!

For (1), a linked list was my first thought too, though given it's during garbage collection it's a slightly tougher nut to crack. A few concepts:
- The resweep logic used for GC stack overflows could be extended to remediate this case at quite the low size penalty, so that if the scheduler stack fills up it'll run through and MARK all objects that still need to be finalized, run a resweep, and finish there. (A perhaps-pathological way to interpret "conservative" garbage collection, but it would at least eventually get it all over multiple gc calls.)
- There's a theoretical blueprint I've been working through for how to just flat _enable allocating memory from hard IRQs_, though I'm not confident that this theory will work out well enough to be worth a full attempt at implementation yet.
- As only the third idea I thought of, `_weakref_entry` could also recycle its own existing linked list pointer and coordinate to share a single schedule call via a suitable classvar.

For (2), not a sematic detail I was aware of when I wrote this implementation! As you say, easy enough to fix. I had an early version of this that actually _did_ do that by inheriting `_weakref_entry` from `_weakref.ref`, I'll have to wind that back.

And as you say, (3) is a fundamental part of what it means for garbage collection to be conservative. My own implementation of `__del__` ran headfirst into this fact when it came to its tests --- see #18005 for details on that, and if you've not done large-scale stochastic testing on #18822, I'd think it quite likely to _also_ have the issue I discovered there with uninitialized padding between stack frames sometimes retaining pointer values for unexpectedly long times.
I can, though, empirically say that MicroPython GC doesn't retain more than two spuriously-uncollected objects across multiple stack/heap scrubbing passes, and that the `-fstrub` GCC option that would eliminate this issue is a much worse performance pessimization than leaving it.

In any case, I'll see about polishing up (1) and (2).

**Comment by dpgeorge** [comment]
I think point (3) is quite a fundamental limitation, and I don't want you to spend/waste your time fixing the other issues if this approach will never be able to work correctly (unless of course you want to do it for the enjoyment of it, which I can fully appreciate!).

The approach in #18822 will always call the callback when the user object is GC'd, and never leave stray `ref`s around (although due to conservative GC it may fail to collect an unused user object, but that's a separate issue, the point is that the weak ref objects are always in a consistent state).

**Comment by AJMansfield** [comment]
> I think point (3) is quite a fundamental limitation, and I don't want you to spend/waste your time fixing the other issues if this approach will never be able to work correctly (unless of course you want to do it for the enjoyment of it, which I can fully appreciate!).

So after thinking through and remembering my design intent a bit more carefully, I can say that while the spurious-reachability thing might be a concern if there's a very expensive resource that you can't afford to not free, spurious reachability can't create the use-after-free hazard you're imagining --- that's actually the main motivation I had for ensuring my `__del__` implementation was _exactly_ correct with regards to the whole "safe disposal of cyclic isolates" part of the spec. (It's also why the volatile double-check in the `_weakref.ref` primitive, to ensure this can't happen even if there's a race that invokes GC while the function is holding only a copy of the weak pointer on the stack.)

The line in `_weakref_entry` that assigns its `next` pointer to point to the object itself is key, because that ensures that the linked list that grows out as more entries are inserted is a _cyclic_ linked list. The "safe disposal of cyclic isolates" guarantees require that all objects in a cyclic isolate be `__del__`-finalized before any of them can be legally freed, which in this case means the object together with all of its `_weakref_entry`s.

**Comment by dpgeorge** [comment]
> So after thinking through and remembering my design intent a bit more carefully, I can say that while the spurious-reachability thing might be a concern if there's a very expensive resource that you can't afford to not free, spurious reachability can't create the use-after-free hazard you're imagining

Ah, yes, you are correct!  I missed that, even though I drew it in the diagram above, the pointer from `_weakref_entry` back to the user object.

So my point (3) is invalid.

I think then you can make this approach pass all the tests I added in #18822.  It would be interesting to see how this compares to that in terms of memory overhead for each `ref`.

**Comment by AJMansfield** [comment]
@dpgeorge I'm reasonably confident that the implementation I have now is algorithmically correct --- though I'm struggling with the tests still. The forest of objects this version of weakref creates is more vulnerable to spurious reachability problems; this will probably need to be tested with a strategy similar to what I used for testing my implementation of `__del__`.

I've modified the object graph topology here to use a much shallower `list`-based `__weakref__`. It's more memory for the single-weakref case, but it avoids pessimizing the number of GC resweeps needed to collect when making a large number of weakrefs to the same object.

For the sake of getting the semantics of `ref` and `finalize` exactly correct without excessive use of `super()` (that being one of the slowest ways to look up a method, based on testing I did a few months ago), I've also moved to inheriting `_weakref.ref` to a much-more-basic `baseref`, and then just wrap rather than inherit it in `weakref.ref` and `finalize`. Possibly, the behavior of `baseref` in calling its target's `__callback__` method could even be moved to modweakref.c, to make this PR independent of user-class del. (i.e. since native classes already have the ability to alloc themselves with FTB set.)

**Comment by dpgeorge** [comment]
> though I'm struggling with the tests still

You should be able to fix the powerpc build error.

For the coverage build failing on `weakref_finalize_collect.py`: the very first part of that test is failing and it looks strange: `f.alive` is `True` and `f.peek()` returns non-`None`, yet the call to `f.detach()` does return `None` which indicates that it's not actually alive??

Also, you may want to freeze in the relevant Python code for the stm32 and qemu ports (to get the code size comparison for them).

File: extmod/modweakref.c [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    { MP_ROM_QSTR(MP_QSTR___call__), MP_ROM_PTR(&weakref_ref_call_obj) }, // allow super().__call__()
+    { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&weakref_ref_del_obj) },
+    { MP_ROM_QSTR(MP_QSTR___callback__), MP_ROM_PTR(&weakref_ref_callback_obj) },
+};
+static MP_DEFINE_CONST_DICT(weakref_ref_locals_dict, weakref_ref_locals_dict_table);
+
+static MP_DEFINE_CONST_OBJ_TYPE(
+    mp_type_weakref_ref,
+    MP_QSTR_ref,
+    MP_TYPE_FLAG_NONE,
```
Comment (AJMansfield) [inline extmod/modweakref.c]: In the presence of #18005, this would want to be `MP_TYPE_FLAG_HAS_FINALISER | MP_TYPE_FLAG_INSTANCE_TYPE` in order to trigger a [Clause 3 resweep](https://peps.python.org/pep-0442/#disposal-of-cyclic-isolates).

A ref could theoretically be marked for collection in the same GC generation as its referent, making the possibility that the callback of a `weakref.ref` could re-attach it to the object graph into a use-after-free hazard.

Potentially, this could be solved either by having `weakref_ref_del` directly invoke `gc_mark_subtree` on its target after it finishes running the target's `__callback__`... or a somewhat complicated version involving a handful of global weakref tables... or maybe trying to make this independent of `__del__` was already a mistake lol.

**Review by AJMansfield — COMMENTED** [review]
Just refactored this to rely on a slightly-more-powerful C-level `_weakref.ref` primitive that allows `weakref.ref` and `weakref._target` to inherit it directly in order to simplify the object graph, and as a windfall benefit make this implementation _mostly_ independent of #18005, except for the post-finaliser resweep additions.

**Comment by andrewleech** [comment]
Should this be closed now with https://github.com/micropython/micropython/pull/18822 being merged?
